File: client/projector-side-drawer/src/main/java/org/teamapps/projector.component/sidedrawer/SideDrawer.java
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */
package org.teamapps.projector.component.sidedrawer;

import org.teamapps.projector.common.format.Color;
import org.teamapps.projector.component.AbstractComponent;
import org.teamapps.projector.component.Component;
import org.teamapps.projector.component.DtoComponentConfig;
import org.teamapps.projector.event.ProjectorEvent;

public class SideDrawer extends AbstractComponent implements DtoSideDrawerEventHandler {

	private final DtoSideDrawerClientObjectChannel clientObjectChannel = new DtoSideDrawerClientObjectChannel(getClientObjectChannel());

	public final ProjectorEvent<Boolean> onExpandedOrCollapsed = new ProjectorEvent<>(clientObjectChannel::toggleExpandedOrCollapsedEvent);

	private final Component containerComponent;
	private Component contentComponent;
	private int width = -1;
	private int height = -1;
	private int marginX;
	private int marginY;
	private DrawerPosition position;
	private Color backgroundColor = null;
	private Color expanderHandleColor = null;

	private boolean collapsible;
	private boolean expanded;


	public SideDrawer(Component containerComponent, Component contentComponent) {
		this.containerComponent = containerComponent;
		this.contentComponent = contentComponent;
	}

	@Override
	public DtoComponentConfig createDto() {
		DtoSideDrawer ui = new DtoSideDrawer();
		mapAbstractConfigProperties(ui);
		ui.setContainerComponent(containerComponent);
		ui.setContentComponent(contentComponent);
		ui.setWidth(width);
		ui.setHeight(height);
		ui.setMarginX(marginX);
		ui.setMarginY(marginY);
		ui.setPosition(position);
		ui.setBackgroundColor(backgroundColor != null ? backgroundColor.toHtmlColorString() : null);
		ui.setExpanderHandleColor(expanderHandleColor != null ? expanderHandleColor.toHtmlColorString() : null);
		ui.setCollapsible(collapsible);
		ui.setExpanded(expanded);
		return ui;
	}

	@Override
	public void handleExpandedOrCollapsed(boolean expanded) {
		onExpandedOrCollapsed.fire(expanded);
	}

	public int getWidth() {
		return width;
	}

	public void setWidth(int width) {
		this.width = width;
		clientObjectChannel.setDimensions(width, height);
	}

	public int getHeight() {
		return height;
	}

	public void setHeight(int height) {
		this.height = height;
		clientObjectChannel.setDimensions(width, height);
	}

	public int getMarginX() {
		return marginX;
	}

	public void setMarginX(int marginX) {
		this.marginX = marginX;
		clientObjectChannel.setMargins(marginX, marginY);
	}

	public int getMarginY() {
		return marginY;
	}

	public void setMarginY(int marginY) {
		this.marginY = marginY;
		clientObjectChannel.setMargins(marginX, marginY);
	}

	public DrawerPosition getPosition() {
		return position;
	}

	public void setPosition(DrawerPosition position) {
		this.position = position;
		clientObjectChannel.setPosition(position);
	}

	public Color getBackgroundColor() {
		return backgroundColor;
	}

	public void setBackgroundColor(Color backgroundColor) {
		this.backgroundColor = backgroundColor;
		clientObjectChannel.setBackgroundColor(backgroundColor != null ? backgroundColor.toHtmlColorString() : null);
	}

	public Color getExpanderHandleColor() {
		return expanderHandleColor;
	}

	public void setExpanderHandleColor(Color expanderHandleColor) {
		this.expanderHandleColor = expanderHandleColor;
		clientObjectChannel.setExpanderHandleColor(expanderHandleColor != null ? expanderHandleColor.toHtmlColorString() : null);
	}

	public boolean isCollapsible() {
		return collapsible;
	}

	public void setCollapsible(boolean collapsible) {
		this.collapsible = collapsible;
	}

	public boolean isExpanded() {
		return expanded;
	}

	public void setExpanded(boolean expanded) {
		this.expanded = expanded;
		clientObjectChannel.setExpanded(expanded);
	}

	public Component getContentComponent() {
		return contentComponent;
	}

	public void setContentComponent(Component contentComponent) {
		this.contentComponent = contentComponent;
		clientObjectChannel.setContentComponent(contentComponent);
	}

}

File: client/projector-side-drawer/src/main/ts/SideDrawer.ts
/*-
 * ========================LICENSE_START=================================
 * TeamApps
 * ---
 * Copyright (C) 2014 - 2022 TeamApps.org
 * ---
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =========================LICENSE_END==================================
 */

import {
	AbstractComponent,
	type Component,
	parseHtml,
	prependChild,
	ProjectorEvent,
	removeClassesByFunction,
	type ServerObjectChannel
} from "projector-client-object-api";
import {
	type DrawerPosition,
	DrawerPositions,
	type DtoSideDrawer,
	type DtoSideDrawer_ExpandedOrCollapsedEvent,
	type DtoSideDrawerCommandHandler,
	type DtoSideDrawerEventSource
} from "./generated";

export class SideDrawer extends AbstractComponent<DtoSideDrawer> implements DtoSideDrawerCommandHandler, DtoSideDrawerEventSource {

	public readonly onExpandedOrCollapsed: ProjectorEvent<DtoSideDrawer_ExpandedOrCollapsedEvent> = new ProjectorEvent();

	private containerComponent: Component;
	private contentComponent: Component | null = null;
	private $main: HTMLElement;

	private $expanderHandle: HTMLElement;

	constructor(config: DtoSideDrawer, serverObjectChannel: ServerObjectChannel) {
		super(config, serverObjectChannel);
		this.containerComponent = config.containerComponent as Component;

		this.$main = parseHtml(`<div class="SideDrawer"></div>`);

		this.setContentComponent(config.contentComponent);
		this.$expanderHandle = parseHtml(`<div class="expander-handle"></div>`);
		this.$expanderHandle.addEventListener("click", () => {
			this.setExpanded(!this.config.expanded);
			this.onExpandedOrCollapsed.fire({expanded: this.config.expanded});
		});
		this.$main.appendChild(this.$expanderHandle);

		prependChild(this.containerComponent.getMainElement(), this.getMainElement());

		this.setBackgroundColor(config.backgroundColor ?? null);
		this.setExpanderHandleColor(config.expanderHandleColor ?? null);

		const resizeObserver = new ResizeObserver(_ => {
			this.updateFloatingPosition();
		});
		resizeObserver.observe(this.containerComponent.getMainElement());

		this.$main.addEventListener("click", ev => ev.stopPropagation());
		this.$main.addEventListener("pointerdown", ev => ev.stopPropagation());
		this.$main.addEventListener("mousedown", ev => ev.stopPropagation());
		this.$main.addEventListener("mouseclick", ev => ev.stopPropagation());
		this.$main.addEventListener("dblclick", ev => ev.stopPropagation());
		this.$main.addEventListener("scroll", ev => ev.stopPropagation());
		this.$main.addEventListener("wheel", ev => ev.stopPropagation());
		this.$main.addEventListener("keydown", ev => ev.stopPropagation());
		this.$main.addEventListener("keyup", ev => ev.stopPropagation());
		this.$main.addEventListener("keypress", ev => ev.stopPropagation());

		this.updateFloatingPosition();
	}

	private updateFloatingPosition() {
		removeClassesByFunction(this.$expanderHandle.classList, className => className.startsWith("position-"));
		this.$expanderHandle.classList.add("position-" + this.config.position);

		let containerWidth = this.containerComponent.getMainElement().offsetWidth;
		let containerHeight = this.containerComponent.getMainElement().offsetHeight;

		// width
		if (this.config.width === -1) {
			this.getMainElement().style.width = '';
			this.getMainElement().style.maxWidth = (containerWidth - 2 * this.config.marginX) + "px";
		} else if (this.config.width === 0) {
			this.getMainElement().style.width = (containerWidth - 2 * this.config.marginX) + "px";
			this.getMainElement().style.maxWidth = '';
		} else {
			this.getMainElement().style.width = this.config.width + "px";
			this.getMainElement().style.maxWidth = '';
		}

		let contentWidth = this.contentComponent?.getMainElement()?.offsetWidth ?? 0;
		// let contentHeight = this.contentComponent?.getMainElement()?.offsetHeight ?? 0;

		// position X
		if (this.config.expanded) {
			if (this.config.position === DrawerPositions.TOP_LEFT || this.config.position === DrawerPositions.BOTTOM_LEFT) {
				this.getMainElement().style.left = this.config.marginX + "px";
				this.getMainElement().style.right = '';
			} else if (this.config.position === DrawerPositions.TOP_RIGHT || this.config.position === DrawerPositions.BOTTOM_RIGHT) {
				this.getMainElement().style.left = '';
				this.getMainElement().style.right = this.config.marginX + "px";
			}
		} else {
			if (this.config.position === DrawerPositions.TOP_LEFT || this.config.position === DrawerPositions.BOTTOM_LEFT) {
				this.getMainElement().style.left = `-${contentWidth}px`;
				this.getMainElement().style.right = '';
			} else if (this.config.position === DrawerPositions.TOP_RIGHT || this.config.position === DrawerPositions.BOTTOM_RIGHT) {
				this.getMainElement().style.left = '';
				this.getMainElement().style.right = `-${contentWidth}px`;
			}
		}

		// position Y
		if (this.config.position === DrawerPositions.TOP_LEFT || this.config.position === DrawerPositions.TOP_RIGHT) {
			this.getMainElement().style.top = this.config.marginY + "px";
			this.getMainElement().style.bottom = '';
		} else if (this.config.position === DrawerPositions.BOTTOM_LEFT || this.config.position === DrawerPositions.BOTTOM_RIGHT) {
			this.getMainElement().style.top = '';
			this.getMainElement().style.bottom = this.config.marginY + "px";
		}
		// height
		if (this.config.height === -1) {
			this.getMainElement().style.height = '';
			this.getMainElement().style.maxHeight = (containerHeight - 2 * this.config.marginY) + "px";
		} else if (this.config.height === 0) {
			this.getMainElement().style.height = (containerHeight - 2 * this.config.marginY) + "px";
			this.getMainElement().style.maxHeight = '';
		} else {
			this.getMainElement().style.height = this.config.height + "px";
			this.getMainElement().style.maxHeight = (containerHeight - 2 * this.config.marginY) + "px";
		}
	}


	doGetMainElement(): HTMLElement {
		return this.$main;
	}

	public setContentComponent(contentComponent: unknown) {
		this.$main.innerHTML = '';
		this.contentComponent = contentComponent as Component;
		if (contentComponent != null) {
			this.$main.appendChild(this.contentComponent.getMainElement());
		}
	}

	setExpanded(expanded: boolean): void {
		this.config.expanded = expanded;
		this.$main.classList.toggle("expanded", expanded);
		this.updateFloatingPosition();
	}

	setPosition(position: DrawerPosition): void {
		this.config.position = position;
		this.updateFloatingPosition();
	}

	setDimensions(width: number, height: number): void {
		this.config.width = width;
		this.config.height = height;
		this.updateFloatingPosition();
	}

	setMargins(marginX: number, marginY: number): void {
		this.config.marginX = marginX;
		this.config.marginY = marginY;
		this.updateFloatingPosition();
	}

	setBackgroundColor(backgroundColor: string | null) {
		this.getMainElement().style.backgroundColor = backgroundColor ?? '';
	}

	setExpanderHandleColor(expanderHandleColor: string | null) {
		this.$expanderHandle.style.backgroundColor = expanderHandleColor ?? '';
	}
}


